src/mutator_test.cc

```diff
@@ -577,6 +578,75 @@ TYPED_TEST(MutatorTypedTest, FailedMutations) {
   EXPECT_LT(crossovers, 100u);
 }
 
+TYPED_TEST(MutatorTypedTest, FieldMutator) {
+  constexpr int kNoMutation = 1;
+  constexpr int kRegularMutation = 2;
+  constexpr int kCustomMutation = 4;
+  constexpr int kAllCases = kNoMutation | kRegularMutation | kCustomMutation;
+
+  constexpr char kIndicatorString[] = "0123456789abcdef";
+
+  TestMutator mutator(false);
+  typename TestFixture::Message message;
+  TestMutator::RegisterCustomMutation(
+      message.GetDescriptor()->FindFieldByName("optional_string"),
+      [kIndicatorString](protobuf::Message* message){
+        typename TestFixture::Message* test_message =
+            dynamic_cast<typename TestFixture::Message*>(message);
+        test_message->set_optional_string(kIndicatorString);
+      });
+  TestMutator::RegisterCustomMutation(
+      message.GetDescriptor()->FindFieldByName("optional_bytes"),
+      [kIndicatorString](protobuf::Message* message){
+        typename TestFixture::Message* test_message =
+            dynamic_cast<typename TestFixture::Message*>(message);
+        test_message->set_optional_bytes(kIndicatorString);
+      });
+
+  int cases_found = 0;
+  for (int j = 0; j < 100000; ++j) {
+    // Include these fields to increase the probability of mutation.
+    if (message.optional_bytes().empty()) {
+      message.set_optional_bytes(" ");
+    }
+    if (message.optional_string().empty()) {
+      message.set_optional_string(" ");
+    }
+    std::string before_bytes = message.optional_bytes();
+    std::string before_string = message.optional_string();
+    mutator.Mutate(&message, 1);
+
+    if (message.optional_bytes() == before_bytes) {
+      cases_found |= kNoMutation;
+    } else if (message.optional_bytes() == kIndicatorString) {
+      cases_found |= kCustomMutation;
+      // Revert the custom mutation to avoid the kCustomMutation case being
+      // counted as the kNoMutation case.
+      message.set_optional_bytes(before_bytes);
+    } else {
+      cases_found |= kRegularMutation;
+    }
+
+    if (message.optional_string() == before_string) {
+      cases_found |= kNoMutation;
+    } else if (message.optional_string() == kIndicatorString) {
+      cases_found |= kCustomMutation;
+      // Revert the custom mutation to avoid the kCustomMutation case being
+      // counted as the kNoMutation case.
+      message.set_optional_string(before_string);
+    } else {
+      cases_found |= kRegularMutation;
+    }
+
+    if (cases_found == kAllCases)
+      break;
+  }
+
+  EXPECT_EQ(cases_found & kNoMutation, kNoMutation);
```

Comment:
Now I think kNoMutation probably not helpful, kRegularMutation covers needed case
Also there is no need to use bit arithmetics here. Bools will make code simpler

Reply:
Done

Reply:
Try move message variable inside the loop so it reset in each iteration.
As is it's creates nested message continuously reducing probability to hit the field.

Reply:
Thanks, that make it significantly faster.